### ISSUE TITLE
[API Proposal]: UInt64.DivRemByUInt32

### ISSUE THREAD

**Issue by lilinus** [dotnet/runtime#98405]
### Background and motivation

There are multiple places in the runtime where we do a DivRem. If we can't get the quotient and remainder in one go using intrinsics or something else, then it is faster to compute the 32-bit remainder (guaranteed since divisor is 32-bit), compared to a 64-bit remainder.

Examples of such places:
Decimal.DecCalc (one example from file)
https://github.com/dotnet/runtime/blob/d4643138e8b9f23482618a32535cc40552da0bb7/src/libraries/System.Private.CoreLib/src/System/Decimal.DecCalc.cs#L215-L217

Uint64 (formatting)
https://github.com/dotnet/runtime/blob/d4643138e8b9f23482618a32535cc40552da0bb7/src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs#L2338-L2339

BigInteger 
https://github.com/dotnet/runtime/blob/d4643138e8b9f23482618a32535cc40552da0bb7/src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.DivRem.cs#L20-L29
DateTime (for decomposing into years, months, days, hours etc).
https://github.com/dotnet/runtime/blob/d4643138e8b9f23482618a32535cc40552da0bb7/src/libraries/System.Private.CoreLib/src/System/DateTime.cs#L1419-L1424

It should be possible to leverage hardware intrisics:
- If we can use intrisicts (like in `DivRem(ulong, ulong)`), we can just truncate the remainder to uint.
- Otherwise remainder can be calculated as `(uint)dividend - ((uint)quotient * divisor)` since 32 bit multiplication should be faster than 64-bit. This is of course preceeded by a 64-bit division.

Haven't seen any use of the same DivRem for signed numbers yet so excluded them from this suggestion

### API Proposal

```csharp
namespace System;

public readonly struct UInt64
{
    public static (ulong Quotient, uint Remainder) DivRemByUInt32(ulong left, uint right);
}
```

### API Usage

```csharp
//
ulong dividend = ulong.MaxValue;
uint divisor = 10;
(ulong quotient, uint remainder) = ulong.DivRemByUInt32(divident, divisor)
```

**Comment by Daniel-Svensson** [comment]
For decimal you probably don't want such a general api, but instead want to use the x86base instrinct, which works only when upper 32bits < divisor which is known to be the case for the majority of the instructions. (I did some modifications back when the code was in c++ and got it 300% faster for financial calculations using just that simple approach togheter with som add/sub with carry).

Hardware division performance has improved since then, but sticking with 32bit divide (64 by 32) instead of 64 bit (128 by 64) is still probably faster than what the code does today. (2 32bit instructions was faster than one 64bit on sandy bridge) and is supported in both 32 and 64bit mode.

**Comment by lilinus** [comment]
Yes, I was thinking it would just be a faster alternative to use when calling DivRem with ulong and uint, and expecting to get a uint remainder (and ulong quotient). This can be consumed in various places, such as those I pointed out in proposal

**Comment by tannergooding** [comment]
This is something we'd rather add pattern recognition in the JIT. Namely `ulong / uint` should be recognized and just do the right thing.
